File: account/src/main/java/serp/project/account/ui/controller/OrganizationController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import serp.project.account.core.domain.constant.Constants;
import serp.project.account.core.domain.dto.request.CreateUserForOrgRequest;
import serp.project.account.core.domain.dto.request.GetOrganizationParams;
import serp.project.account.core.domain.dto.request.UpdateUserStatusRequest;
import serp.project.account.core.usecase.OrganizationUseCase;
import serp.project.account.core.usecase.UserUseCase;
import serp.project.account.kernel.utils.AuthUtils;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1")
public class OrganizationController {
    private final OrganizationUseCase organizationUseCase;
    private final UserUseCase userUseCase;

    private final AuthUtils authUtils;

    @GetMapping("/admin/organizations")
    public ResponseEntity<?> getOrganizations(
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String status,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer pageSize) {
        GetOrganizationParams params = GetOrganizationParams.builder()
                .search(search)
                .status(status)
                .type(type)
                .page(page)
                .pageSize(pageSize)
                .build();
        var response = organizationUseCase.getOrganizations(params);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/admin/organizations/{organizationId}")
    public ResponseEntity<?> getOrganizationById(@RequestParam Long organizationId) {
        var response = organizationUseCase.getOrganizationById(organizationId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/organizations/me")
    public ResponseEntity<?> getMyOrganization() {
        Long organizationId = authUtils.getCurrentTenantId().orElse(null);
        var response = organizationUseCase.getOrganizationById(organizationId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/organizations/{organizationId}/users")
    public ResponseEntity<?> createUserForOrganization(
            @PathVariable Long organizationId,
            @Valid @RequestBody CreateUserForOrgRequest request

    ) {
        if (!authUtils.canAccessOrganization(organizationId)) {
            return ResponseEntity.status(403).body(Constants.ErrorMessage.FORBIDDEN);
        }
        var response = organizationUseCase.createUserForOrganization(organizationId, request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/organizations/{organizationId}/users/{userId}/status")
    public ResponseEntity<?> updateUserStatusInOrganization(
            @PathVariable Long organizationId,
            @PathVariable Long userId,
            @Valid @RequestBody UpdateUserStatusRequest request) {
        boolean isSerpAdmin = authUtils.isSystemAdmin();
        Long updatedBy = authUtils.getCurrentUserId().orElse(null);
        var response = userUseCase.updateUserStatus(
                organizationId,
                updatedBy,
                userId,
                request.getStatus(),
                isSerpAdmin);
        return ResponseEntity.status(response.getCode()).body(response);
    }

}

File: api_gateway/src/ui/controller/account/organization_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type OrganizationController struct {
	organizationService service.IOrganizationService
}

func (o *OrganizationController) GetOrganizations(c *gin.Context) {
	page, pageSize, valid := utils.ValidatePaginationParams(c)
	if !valid {
		return
	}

	search := utils.ParseStringQuery(c, "search")
	status := utils.ParseStringQuery(c, "status")
	organizationType := utils.ParseStringQuery(c, "type")

	params := &request.GetOrganizationParams{
		Search:           search,
		Status:           status,
		OrganizationType: organizationType,
		Page:             &page,
		PageSize:         &pageSize,
	}

	res, err := o.organizationService.GetOrganizations(c.Request.Context(), params)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (o *OrganizationController) GetOrganizationById(c *gin.Context) {
	organizationID, err := utils.ParseInt64Param(c, "organizationId")
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := o.organizationService.GetOrganizationById(c.Request.Context(), organizationID)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (o *OrganizationController) GetMyOrganization(c *gin.Context) {
	res, err := o.organizationService.GetMyOrganization(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (o *OrganizationController) CreateUserForOrganization(c *gin.Context) {
	var req request.CreateUserForOrgRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	organizationID, valid := utils.ValidateAndParseID(c, "organizationId")
	if !valid {
		return
	}

	res, err := o.organizationService.CreateUserForOrganization(c.Request.Context(), organizationID, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}

	c.JSON(res.Code, res)
}

func NewOrganizationController(organizationService service.IOrganizationService) *OrganizationController {
	return &OrganizationController{
		organizationService: organizationService,
	}
}
